### PR TITLE
Implement `allow_missing` for JWT requirement.

### PR DESCRIPTION
**Pull request by diemtvu**
Description:
Implement `allow_missing` (but not failed) mode for JWT requirement. In this mode, token is not required for JWT filter to pass. However, if a token is presented, it must be valid.

This PR also change the `allow_missing_or_failed` (and the new `allow_missing`) to consider the provider within the container OR-list, instead of using the global one that have all providers. As a result, each `allow_xxx` will examine the interested tokens list independently. This correct the behavior when the `allow_xxx` is used multiple times, e.g in different ORs of an AND-of-OR requirements.

Risk Level: Medium

The PR changes the existing behavior of `allow_missing_or_failed`

Testing:
Unit tests.
Docs Changes:
Release Notes:
[Optional Fixes #Issue]
https://github.com/envoyproxy/envoy/issues/8909
https://github.com/istio/istio/issues/18399
[Optional Deprecated:]

### CODE REVIEW

File: api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto [hunk 1]
```diff
@@ -279,12 +291,17 @@ message JwtRequirement {
     // All of them must pass, if one of them fails or missing, it fails.
     JwtRequirementAndList requires_all = 4;
 
+    // Set both allow_missing and allow_invalid instead to have the same effect.
```
Comment (htuch) [inline api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto]: Should this field now be tagged as `deprecated`?

Comment (diemtvu) [reply api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto]: I reverted this. I think it's better to keep the current `allow_missing_or_failed` and add `allow_missing` (but not failed). I don't think the mode `allow_failed` make any sense by itself, so keeping `allow_missing_or_failed` is better than deprecate it and replace with `allow_missing` and `allow_failed` as suggested previously in the related [issue](https://github.com/envoyproxy/envoy/issues/8909).

File: api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto [hunk 2]
```diff
@@ -262,7 +262,19 @@ message ProviderWithAudiences {
...
 //          - provider_name: provider-C
 //
-// [#next-free-field: 6]
+//  # Example 7: A is optional (if token from A is provided, it must be valid, but also allows missing token.)
+//  requirements:
+//  - provider_name: provider-A
+//    allow_missing: true
+//
+//  # Example 8: A is optional and B is required.
+//  requires_all:
```
Comment (lizan) [inline api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto]: can you update this example to reflect last change?

Comment (diemtvu) [reply api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto]: Done

**Comment by diemtvu** [comment]
@lizan gentle ping

**Comment by diemtvu** [comment]
@lizan, can you take a look. Thanks

**Comment by diemtvu** [comment]
How can I get this move forward?

**Comment by lizan** [comment]
sorry for the delay, this is good to move forward now, but in long term we should simplify the JWT API quite a bit.

**Comment by lizan** [comment]
@diemtvu can you add a note to version_history?

**Comment by diemtvu** [comment]
> @diemtvu can you add a note to version_history?

done

**Comment by lizan** [comment]
@diemtvu can you fix CI?

**Comment by diemtvu** [comment]
> @diemtvu can you fix CI?

done

File: docs/root/intro/version_history.rst [hunk 3]
```diff
@@ -31,6 +31,8 @@ Version history
 * tls: remove TLS 1.0 and 1.1 from client defaults
 * tracing: added the ability to set custom tags on both the :ref:`HTTP connection manager<envoy_api_field_config.filter.network.http_connection_manager.v2.HttpConnectionManager.tracing>` and the :ref:`HTTP route <envoy_api_field_route.Route.tracing>`.
 * udp: added initial support for :ref:`UDP proxy <config_udp_listener_filters_udp_proxy>`
+* router: exposed DOWNSTREAM_REMOTE_ADDRESS as custom HTTP request/response headers.
+* jwt_authn filter: added `allow_missing` option
```
Comment (mattklein123) [inline docs/root/intro/version_history.rst]: Merge issue here with the router entry, please also alpha order the new entry.

Comment (diemtvu) [reply docs/root/intro/version_history.rst]: Done

**Comment by diemtvu** [comment]
Ping again to help merging

**Pull request merged**